### PR TITLE
KOA-6021 Update NavBarfor MD3 – Compose

### PR DESCRIPTION
**Pull request by bvitaliyg**
+ [ ] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-android/blob/main/CONTRIBUTING.md)

Remember to include the following changes:
+ [ ] Component `README.md`
+ [ ] Tests

If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/main/CODE_REVIEW_GUIDELINES.md)

### CODE REVIEW

File: app/src/main/java/net/skyscanner/backpack/demo/compose/NavBarStory.kt [hunk 1]
```diff
@@ -121,29 +120,27 @@ fun TransparentNavBarStory(
         Box(modifier = Modifier.padding(vertical = BpkSpacing.Base)) {
             Image(
                 painter = painterResource(id = R.drawable.swimming),
-                modifier = Modifier.fillMaxWidth(),
+                modifier = Modifier.matchParentSize(),
                 contentDescription = "",
                 contentScale = ContentScale.FillWidth,
             )
-            Column(modifier = Modifier.align(Alignment.BottomEnd)) {
```
Comment (bvitaliyg) [inline app/src/main/java/net/skyscanner/backpack/demo/compose/NavBarStory.kt:128]: The story was behaving wrong and weirdly, I made it close to the real use case.

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/navigationbar/BpkTopNavBar.kt [hunk 2]
```diff
@@ -41,8 +41,16 @@ internal sealed interface Action {
     val onClick: () -> Unit
 }
 
-data class IconAction(val icon: BpkIcon, val contentDescription: String, override val onClick: () -> Unit) : Action
-data class TextAction(val text: String, override val onClick: () -> Unit) : Action
+data class IconAction(
```
Comment (bvitaliyg) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/navigationbar/BpkTopNavBar.kt:44]: Just some code formatting

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/navigationbar/BpkTopNavBar.kt [hunk 3]
```diff
@@ -51,14 +59,16 @@ fun BpkTopNavBar(
     modifier: Modifier = Modifier,
     insets: WindowInsets? = WindowInsets.statusBars,
     actions: List<IconAction> = emptyList(),
+    style: NavBarStyle = NavBarStyle.Default,
```
Comment (bvitaliyg) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/navigationbar/BpkTopNavBar.kt:62]: API alignment – this wasn't included to fixed TopNavBar

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/navigationbar/TopNavBarState.kt [hunk 4]
```diff
@@ -41,25 +39,46 @@ fun Modifier.nestedScroll(state: TopNavBarState): Modifier =
...
-    val flingBehavior = ScrollableDefaults.flingBehavior()
-    return rememberSaveable(
-        offsetRange, flingBehavior,
-        saver = TopNavBarStateImpl.saver(offsetRange, flingBehavior),
-        init = { TopNavBarStateImpl(initialStatus, offsetRange = offsetRange, flingBehavior = flingBehavior) },
+}
+
+@Composable
+@OptIn(ExperimentalMaterial3Api::class)
+internal fun rememberFixedTopAppBarState(): TopNavBarState {
```
Comment (bvitaliyg) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/navigationbar/TopNavBarState.kt:64]: Internal usage only: this is for the toolbar without any collapsing effect

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/navigationbar/TopNavBarState.kt [hunk 5]
```diff
@@ -41,25 +39,46 @@ fun Modifier.nestedScroll(state: TopNavBarState): Modifier =
...
 }
 
 @Stable
-internal interface TopNavBarInternalState : TopNavBarState {
-
-    val fraction: Float
-
-    val nestedScrollConnection: NestedScrollConnection
-}
+@OptIn(ExperimentalMaterial3Api::class)
```
Comment (bvitaliyg) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/navigationbar/TopNavBarState.kt:78]: It was previously one internal interface + one internal implementation. Merged now.

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/navigationbar/internal/BpkTopNavBarImpl.kt [hunk 6]
```diff
@@ -55,21 +49,25 @@ import net.skyscanner.backpack.compose.tokens.NativeAndroidClose
...
-    val backgroundColor = if (fraction == 0f) {
-        BpkTheme.colors.surfaceDefault
-    } else if (style == NavBarStyle.OnImage) {
-        Color.Transparent
-    } else {
-        BpkTheme.colors.canvas
+    val internalState = scrollBehavior.asInternalState()
+    val fraction = 1f - internalState.state.collapsedFraction
+
+    val backgroundColor = when (fraction) {
```
Comment (bvitaliyg) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/navigationbar/internal/BpkTopNavBarImpl.kt:65]: Replaced ifs with when statements

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/navigationbar/internal/TopAppBarLayout.kt [hunk 7]
```diff
@@ -18,100 +18,342 @@
...
+import kotlin.math.abs
+import kotlin.math.max
+import kotlin.math.roundToInt
 
 @Composable
-internal fun TopAppBarLayout(
-    fraction: Float,
-    navIcon: @Composable BoxScope.() -> Unit,
-    title: @Composable BoxScope.() -> Unit,
+@OptIn(ExperimentalMaterial3Api::class)
```
Comment (bvitaliyg) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/navigationbar/internal/TopAppBarLayout.kt:66]: This whole file is easier to review as a brand-new component. It's a fork of MD3 NavBar with some simplified code.

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/navigationbar/internal/TopNavBarState.kt [hunk 8]
```diff
@@ -1,131 +0,0 @@
...
-import androidx.compose.runtime.setValue
-import androidx.compose.ui.geometry.Offset
-import androidx.compose.ui.input.nestedscroll.NestedScrollConnection
-import androidx.compose.ui.input.nestedscroll.NestedScrollSource
-import androidx.compose.ui.unit.Velocity
-import net.skyscanner.backpack.compose.navigationbar.TopNavBarInternalState
-import net.skyscanner.backpack.compose.navigationbar.TopNavBarStatus
-
-@Stable
-internal class TopNavBarStateImpl private constructor(
```
Comment (bvitaliyg) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/navigationbar/internal/TopNavBarState.kt:36]: Not needed anymore – this whole implementation is replaced with the default state delegate

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/navigationbar/BpkTopNavBar.kt (hunk 3 above)
Comment (marianeum) [reply backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/navigationbar/BpkTopNavBar.kt:62]: only the expanded one has the transparent style, so it wasn't added intentionally

**Review by marianeum — COMMENTED** [review]
looking at the screenshot there's something up where the collapsed toolbar is still transparent, although it should be solid

**Review by marianeum — APPROVED** [review]

**Pull request merged**